src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T value) {
+            this.value = value;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> node;
+    private int size;
+
+    public MyLinkedList() {
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        Node current = new Node(value);
+        if (size == 0) {
+            node = current;
+            head = node;
+            tail = node;
+            size++;
+        } else {
+            node.next = current;
+            current.prev = node;
+            tail = current;
+            node = node.next;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node current = new Node(value);
+        if (size == 0) {
+            node = current;
+            head = node;
+            tail = node;
+            size++;
+            return;
+        }
+        if (index == 0) {
+            head.prev = current;
+            current.next = head;
+            head = current;
+            size++;
+            return;
+        }
+        if (index == size) {
+            node.next = current;
+            current.prev = node;
+            tail = current;
+            node = node.next;
+            size++;
+        } else {
+            Node<T> counter = getElementByIndex(index);
+            current.prev = counter.prev;
+            counter.prev.next = current;
+            counter.prev = current;
+            current.next = counter;
+            size++;
+            return;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        return getElementByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> current = getElementByIndex(index);
+        T result = current.value;
+        current.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        T result;
+        if (index == 0 && size == 1) {
```

Comment:
looks like hardcoding

let's call getElementByIndex(index) -> and then unlink(node)